### PR TITLE
Restore RESTEasy codestart by default for Quarkus < 2.8

### PR DESCRIPTION
**Pull request by gsmet**
This is very hackyish and we need to do better but that should do it
until we find a better solution.

@quarkusio/devtools that's what I could come up with as a band aid for now until we resolve this thing properly.
@aloubyansky I'm especially interested in your feedback for the BOM test. I know it's not pretty and we will need to do better but do you think it's good enough to cover most cases for now?

### CODE REVIEW

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java
Comment (maxandersen) [inline independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java:174]: this could possibly work but its really a bad hacky fish (as you said :) - before we +1 this as opposed to change it to the behavior we had before (add resteasy or possibly now resteasy-reactive by default when no codestarts) ... is there something around this that is important for the resteasy-reactive change?

Since I would much rather go with the less optimal but we know works (or at least examples/humans are adjusted to) than this change we risk having to deal with for a long time.

Comment (gsmet) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java:174]: You need the hack anyway if you need to default to RESTEasy Reactive for 2.8+

Comment (gsmet) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java:174]: Obviously we will make it cleaner in 2.8.1 with a proper version check.

Comment (gsmet) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java:174]: FWIW I'm working on a cleaner version.

**Comment by maxandersen** [comment]
One idea is to use platform property to control this. 

Then no version specific check needed.

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java
Comment (gsmet) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java]: Pushed a new version.

As for enabling RESTEasy Reactive by default, again, we want users to select what they want and not force push a REST impl + examples to them.
Restoring this default behavior will probably break tests and it will take quite some time to figure out all of them especially in the Maven/Gradle tests as they take ages to run.
If you want to pursue this, I'll let you find a guinea pig. I strongly believe it's something we shouldn't do, especially with all the different use cases Quarkus now has.

Comment (gsmet) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java]: Note that the simple case where people just create a default app still provides a default RESTEasy Reactive app.
It's when you go choose your extensions that you need to make an explicit choice.

**Comment by gsmet** [comment]
> One idea is to use platform property to control this.
> Then no version specific check needed.

Yes but that's definitely not something that will be done by tomorrow morning.
I think we can live with this workaround for 2.8.0.

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java
Comment (aloubyansky) [inline independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java:45]: It could also be a productized platform, which will have a different groupId.

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java
Comment (aloubyansky) [inline independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java:214]: This is where you could check for `com.redhat.quarkus.platform`

**Comment by ia3andy** [comment]
My take on this is that we should add the default codestart in the platform metadata. For version before 2.8, we could either force it in the registry or use a default to RESTEasy @aloubyansky any thought?

**Comment by aloubyansky** [comment]
What do you mean by forcing it in the registry @ia3andy? Wouldn't we still need a change in the code to support it?

**Comment by ia3andy** [comment]
> Then in this PR, we just have to read it and it will be already version dependent.

Yes, then in this PR, we just have to read it and it will be already version dependent.

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java
Comment (maxandersen) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java:45]: I agree, but for 2.8/2.9 that would (most likely) not apply. To handle that it would be better checking platform property.

Comment (aloubyansky) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java:45]: Right. I suppose it will "break" when using dev tools 2.8 with the next rh release (post 2.7).

**Comment by maxandersen** [comment]
As 2.8.0.Final is building today I don't see us do a better fix; thus +1 on this with the following caveat: 

we get a PR for 2.8.1 ready that use platform property where not present at all will fall back to adding resteasy by default, if present it can set the default code start or if blank it means "dont add any".

This will have the consequence of 2.8.0 platform will not generate the right thing but anything before and after will be controllable; but that is a tolerable bug imo as alternative is we hold up release to do this right.

File: independent-projects/tools/devtools-common/pom.xml
Comment (maxandersen) [inline independent-projects/tools/devtools-common/pom.xml:60]: minor issue that can be fixed later. if this is just for parsing version devtools should already have maven GAV parser available - @aloubyansky ?

**Comment by ia3andy** [comment]
- whatever your change in the code, previous versions of the CLI will create app with RESTEasy classic
- newer version of the CLI will have this code to use a specific default depending on the platform data

it would be best if we could also set this default codestart in previous versions of the platform in the registry. Then in the code if there is no default we don't use any.

**Comment by gsmet** [comment]
You need something more clever than just put a default codestart. Because, really, we don't want to pursue with this "let's force a REST impl on everybody" thing.
I talked with several people in the team and most of them were surprised that this was still in place.

**Review by maxandersen — APPROVED** [review]

File: independent-projects/tools/devtools-common/pom.xml
Comment (aloubyansky) [reply independent-projects/tools/devtools-common/pom.xml:60]: We don't have a "home made" one. There is `org.apache.maven.artifact.versioning.DefaultArtifactVersion` that could have been used for this. I suppose @gsmet didn't want to depend on that API.

Comment (aloubyansky) [reply independent-projects/tools/devtools-common/pom.xml:60]: It comes from `org.apache.maven:maven-artifact` which should already be on CP as a transitive dep.

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java
Comment (gsmet) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java]: I added that.

**Comment by aloubyansky** [comment]
I guess we also need to enhance devmode to include the dev ui by default even if none of the REST impl isn't included in the app.

**Review by aloubyansky — APPROVED** [review]

**Comment by aloubyansky** [comment]
> * whatever your change in the code, previous versions of the CLI will create app with RESTEasy classic
> * newer version of the CLI will have this code to use a specific default depending on the platform data
> 
> it would be best if we could also set this default codestart in previous versions of the platform in the registry. Then in the code if there is no default we don't use any.

That would be problematic unless we make the registry actually provide the platform metadata. Currently the registry provides only non-platform catalogs while the platform ones are pulled from regular public Maven repos.

**Comment by ia3andy** [comment]
> > * whatever your change in the code, previous versions of the CLI will create app with RESTEasy classic
> > * newer version of the CLI will have this code to use a specific default depending on the platform data
> > 
> > it would be best if we could also set this default codestart in previous versions of the platform in the registry. Then in the code if there is no default we don't use any.
> 
> That would be problematic unless we make the registry actually provide the platform metadata. Currently the registry provides only non-platform catalogs while the platform ones are pulled from regular public Maven repos.

Ah ok I was unsure about it, so it means we can only add it for 2.8.0+

**Review by ia3andy — CHANGES_REQUESTED** [review]
We need to fix the RESTEasy classic extensions binding before merging this.

See the discussion on zulip.

**Review by ia3andy — APPROVED** [review]
LGTM (my concern was already fixed: https://github.com/quarkusio/quarkus/pull/24413/commits/1e0457458f7e77845b16a0021ea2f74d8a76a470)

**Comment by gsmet** [comment]
CI had already passed. I just added the RHBQ Platform so I can merge and proceed with the release.

**Pull request merged**
